analyses/cms-open-data-ttbar/validate_histograms.py

```diff
@@ -42,8 +43,57 @@ def validate(histos: dict, reference: dict) -> dict[str, list[str]]:
-            errors[name].append(f"Contents do not match:\n\tgot      {h['contents']}\n\texpected {ref_h['contents']}")
+        is_close = np.isclose(h['contents'], ref_h['contents']) 
+        
+        #### check if bin migration ####
+        if not contents_depend_on_rng and not all(is_close):
+
+            # gets indices where above array is false
+            where_not_close = np.where(np.invert(is_close))[0]
+            # gets difference of adjacent entries
+            diff_values = np.diff(where_not_close)
+            # get the indices where the above difference is greater than 1 (to form groupings of bins based on location)
+            split_indices = np.argwhere(np.abs(diff_values)>1)
+            # np.argwhere adds extra dimension we don't need
+            split_indices = split_indices.reshape((split_indices.shape[0],))
+            
+            # if only one grouping detected
+            if len(split_indices)==0:
+                split_values = [where_not_close]
+            # if more than one grouping detected
+            else:
+                # shift indices to the point we want to split at
+                split_indices = split_indices + 1
+                # split indices into groupings
+                split_values = np.split(where_not_close, split_indices)
+            
+            is_error=False
+            # iterate through groupings and test for bin migration or error
+            for group in split_values:
+                h_group = np.array(h['contents'])[group]
+                ref_group = np.array(ref_h['contents'])[group]
+                # if difference is great, count as error
+                if not np.allclose(h_group, ref_group, atol=2.0): # 2.0 is chosen as it seems to cover the difference expected of one events migrating between bins in histograms from 1 file per sample. this definitely can be tuned in the future
+                    is_error = True
+                    if verbose:
+                        print(f"In {name}: Not close enough for bin migration")
+                        print("histogram: ", h_group, ", reference: ", ref_group)
+                        print()
+                # check partial sum
+                elif not np.allclose(sum(h_group), sum(ref_group)):
+                    is_error = True
+                    if verbose:
+                        print(f"In {name}: Partial sums are unequal")
+                        print("histogram: ", h_group, ", reference: ", ref_group)
+                        print()
+                else:
+                    if verbose:
+                        print(f"In {name}: Bin migration likely")
+                        print("histogram: ", h_group, ", reference: ", ref_group)
+                        print()
+            if is_error:
+                errors[name].append(f"Contents do not match:\n\tgot      {h['contents']}\n\texpected {ref_h['contents']}")
+                print()
```

Comment:
I see we also have it in `main`, but isn't this `print()` completely superfluous? It only fills up the output with blank lines in case of disagreements.

Reply:
I think you are right, let me remove it. I have to fix #176 as well

Reply:
removed

Reply:
Thanks!